File: lemur-libs/lemur-testkit/src/main/java/io/github/malczuuu/lemur/testkit/kafka/TestKafkaConsumerRegistrar.java
package io.github.malczuuu.lemur.testkit.kafka;

import io.github.malczuuu.lemur.testkit.annotation.TestListener;
import java.lang.reflect.Field;
import java.util.HashMap;
import java.util.LinkedHashSet;
import java.util.Map;
import java.util.Set;
import org.apache.kafka.clients.consumer.ConsumerConfig;
import org.apache.kafka.clients.consumer.KafkaConsumer;
import org.apache.kafka.common.serialization.StringDeserializer;
import org.jspecify.annotations.NullUnmarked;
import org.springframework.beans.factory.BeanFactory;
import org.springframework.beans.factory.config.BeanDefinition;
import org.springframework.beans.factory.support.AutowireCandidateQualifier;
import org.springframework.beans.factory.support.BeanDefinitionRegistry;
import org.springframework.beans.factory.support.RootBeanDefinition;
import org.springframework.boot.kafka.autoconfigure.KafkaConnectionDetails;
import org.springframework.core.env.Environment;

/**
 * Scans the test class hierarchy for {@link
 * io.github.malczuuu.lemur.testkit.annotation.TestListener}-annotated fields and registers a {@link
 * TestKafkaConsumer} bean per unique topic, qualified with {@link
 * io.github.malczuuu.lemur.testkit.annotation.TestListener}.
 */
@NullUnmarked
class TestKafkaConsumerRegistrar {

  private final Class<?> testClass;
  private final Environment environment;
  private final BeanFactory beanFactory;

  TestKafkaConsumerRegistrar(Class<?> testClass, Environment environment, BeanFactory beanFactory) {
    this.testClass = testClass;
    this.environment = environment;
    this.beanFactory = beanFactory;
  }

  void registerBeanDefinitions(BeanDefinitionRegistry registry) {
    Class<?> clazz = testClass;
    while (clazz != null && clazz != Object.class) {
      for (Field field : clazz.getDeclaredFields()) {
        TestListener annotation = field.getAnnotation(TestListener.class);
        if (annotation != null) {
          String rawValue = annotation.value();
          String resolvedTopic = environment.resolveRequiredPlaceholders(rawValue);
          String beanName = "testKafkaConsumer[" + rawValue + "]";
          if (!registry.containsBeanDefinition(beanName)) {
            registry.registerBeanDefinition(beanName, buildDefinition(rawValue, resolvedTopic));
          }
        }
      }
      clazz = clazz.getSuperclass();
    }
  }

  private BeanDefinition buildDefinition(String rawValue, String resolvedTopic) {
    RootBeanDefinition definition = new RootBeanDefinition(TestKafkaConsumer.class);
    definition.setInstanceSupplier(
        () -> {
          KafkaConnectionDetails connectionDetails =
              beanFactory.getBean(KafkaConnectionDetails.class);
          String bootstrapServers =
              String.join(",", connectionDetails.getConsumer().getBootstrapServers());
          return createClient(bootstrapServers, resolvedTopic);
        });
    definition.setDestroyMethodNames("close");
    definition.addQualifier(new AutowireCandidateQualifier(TestListener.class, rawValue));
    return definition;
  }

  private TestKafkaConsumer createClient(String bootstrapServers, String resolvedTopic) {
    Map<String, Object> props = new HashMap<>();
    props.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, bootstrapServers);
    props.put(ConsumerConfig.GROUP_ID_CONFIG, "test-consumer." + resolvedTopic);
    props.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
    props.put(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, false);
    KafkaConsumer<String, String> consumer =
        new KafkaConsumer<>(props, new StringDeserializer(), new StringDeserializer());
    return new TestKafkaConsumer(consumer, resolvedTopic);
  }

  static Set<String> collectTopicAnnotatedFieldTypes(Class<?> testClass) {
    Set<String> raw = new LinkedHashSet<>();
    Class<?> clazz = testClass;
    while (clazz != null && clazz != Object.class) {
      for (Field field : clazz.getDeclaredFields()) {
        TestListener annotation = field.getAnnotation(TestListener.class);
        if (annotation != null) {
          raw.add(annotation.value());
        }
      }
      clazz = clazz.getSuperclass();
    }
    return raw;
  }
}

File: lemur-libs/lemur-testkit/src/main/kotlin/io/github/malczuuu/lemur/testkit/kafka/TestKafkaConsumerRegistrar.kt
package io.github.malczuuu.lemur.testkit.kafka

import io.github.malczuuu.lemur.testkit.annotation.TestListener
import org.apache.kafka.clients.consumer.ConsumerConfig
import org.apache.kafka.clients.consumer.KafkaConsumer
import org.apache.kafka.common.serialization.StringDeserializer
import org.springframework.beans.factory.BeanFactory
import org.springframework.beans.factory.support.AutowireCandidateQualifier
import org.springframework.beans.factory.support.BeanDefinitionRegistry
import org.springframework.beans.factory.support.RootBeanDefinition
import org.springframework.boot.kafka.autoconfigure.KafkaConnectionDetails
import org.springframework.core.env.Environment
import java.util.UUID

internal class TestKafkaConsumerRegistrar(
    private val testClass: Class<*>,
    private val environment: Environment,
    private val beanFactory: BeanFactory,
) {

    fun registerBeanDefinitions(registry: BeanDefinitionRegistry) {
        var clazz: Class<*>? = testClass
        while (clazz != null && clazz != Any::class.java) {
            for (field in clazz.declaredFields) {
                val annotation = field.getAnnotation(TestListener::class.java)
                if (annotation != null) {
                    val rawValue = annotation.value
                    val resolvedTopic = environment.resolveRequiredPlaceholders(rawValue)
                    val beanName = "testKafkaConsumer[$rawValue]"
                    if (!registry.containsBeanDefinition(beanName)) {
                        registry.registerBeanDefinition(
                            beanName,
                            buildDefinition(rawValue, resolvedTopic),
                        )
                    }
                }
            }
            clazz = clazz.superclass
        }
    }

    private fun buildDefinition(rawValue: String, resolvedTopic: String): RootBeanDefinition {
        val definition = RootBeanDefinition(TestKafkaConsumer::class.java)
        definition.setInstanceSupplier {
            val connectionDetails = beanFactory.getBean(KafkaConnectionDetails::class.java)
            val bootstrapServers =
                connectionDetails.consumer.bootstrapServers.joinToString(",")
            createClient(bootstrapServers, resolvedTopic)
        }
        definition.setDestroyMethodNames("close")
        definition.addQualifier(AutowireCandidateQualifier(TestListener::class.java, rawValue))
        return definition
    }

    private fun createClient(bootstrapServers: String, resolvedTopic: String): TestKafkaConsumer {
        val props =
            mapOf(
                ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG to bootstrapServers,
                ConsumerConfig.GROUP_ID_CONFIG to getUniqueGroupId(resolvedTopic),
                ConsumerConfig.AUTO_OFFSET_RESET_CONFIG to "earliest",
                ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG to false,
            )
        val consumer = KafkaConsumer(props, StringDeserializer(), StringDeserializer())
        return DefaultTestKafkaConsumer(consumer, resolvedTopic)
    }

    private fun getUniqueGroupId(topic: String): String = "test-consumer.$topic.${UUID.randomUUID()}"
}
